pkg/gadgets/sigsnoop/tracer/tracer.go

```diff
@@ -0,0 +1,279 @@
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	containercollection "github.com/kinvolk/inspektor-gadget/pkg/container-collection"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/sigsnoop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+
+	"golang.org/x/sys/unix"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang sigsnoop ./bpf/sigsnoop.bpf.c -- -I../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Config struct {
+	// TODO: Make it a *ebpf.Map once
+	// https://github.com/cilium/ebpf/issues/515 and
+	// https://github.com/cilium/ebpf/issues/517 are fixed
+	MountnsMap string
+
+	TargetSignal string
+	TargetPid    int32
+	FailedOnly   bool
+}
+
+type Tracer struct {
+	config *Config
+
+	objs               sigsnoopObjects
+	enterKillLink      link.Link
+	exitKillLink       link.Link
+	enterTkillLink     link.Link
+	exitTkillLink      link.Link
+	enterTgkillLink    link.Link
+	exitTgkillLink     link.Link
+	signalGenerateLink link.Link
+	reader             *perf.Reader
+
+	resolver      containercollection.ContainerResolver
+	eventCallback func(types.Event)
+	node          string
+}
+
+func signalStringToInt(signal string) (int32, error) {
+	// There are three possibilities:
+	// 1. Either user did not give a signal, thus the argument is empty string.
+	// 2. Or signal begins with SIG.
+	// 3. Or signal is a string which contains an integer.
+	if signal == "" {
+		return 0, nil
+	}
+
+	if strings.HasPrefix(signal, "SIG") {
+		signalNum := unix.SignalNum(signal)
+		if signalNum == 0 {
+			return 0, fmt.Errorf("no signal found for %q", signal)
+		}
+
+		return int32(signalNum), nil
+	}
+
+	signalNum, err := strconv.ParseInt(signal, 10, 32)
+
+	return int32(signalNum), err
+}
+
+func signalIntToString(signal int) string {
+	return unix.SignalName(syscall.Signal(signal))
+}
+
+func NewTracer(config *Config, resolver containercollection.ContainerResolver,
+	eventCallback func(types.Event), node string) (*Tracer, error) {
+	t := &Tracer{
+		config:        config,
+		resolver:      resolver,
+		eventCallback: eventCallback,
+		node:          node,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Stop() {
+	if t.enterKillLink != nil {
+		t.enterKillLink.Close()
+		t.enterKillLink = nil
+	}
+	if t.exitKillLink != nil {
+		t.exitKillLink.Close()
+		t.exitKillLink = nil
+	}
+
+	if t.enterTkillLink != nil {
+		t.enterTkillLink.Close()
+		t.enterTkillLink = nil
+	}
+	if t.exitTkillLink != nil {
+		t.exitTkillLink.Close()
+		t.exitTkillLink = nil
+	}
+
+	if t.enterTgkillLink != nil {
+		t.enterTgkillLink.Close()
+		t.enterTgkillLink = nil
+	}
+	if t.exitTgkillLink != nil {
+		t.exitTgkillLink.Close()
+		t.exitTgkillLink = nil
+	}
+
+	if t.signalGenerateLink != nil {
+		t.signalGenerateLink.Close()
+		t.signalGenerateLink = nil
+	}
```

Comment:
Please update this with the helper introduced in https://github.com/kinvolk/inspektor-gadget/pull/534